src/firmware/shared/circular_buffer.c

```diff
@@ -0,0 +1,68 @@
+#include "firmware/shared/circular_buffer.h"
+
+#include <assert.h>
+
+struct CircularBuffer
+{
+    int head;
+    int tail;
+    size_t max_size;
+    float *buffer_data;
+    bool isFull;
+};
+
+CircularBuffer_t *circular_buffer_create(size_t size)
+{
+    CircularBuffer_t *new_cbuffer = malloc(sizeof(CircularBuffer_t));
+
+    new_cbuffer->head        = 0;
+    new_cbuffer->tail        = 0;
+    new_cbuffer->max_size    = size;
+    new_cbuffer->buffer_data = (float *)malloc(sizeof(float) * new_cbuffer->max_size);
+    new_cbuffer->isFull      = false;
+
+    return new_cbuffer;
+}
+
+void circular_buffer_destroy(CircularBuffer_t *cbuffer)
+{
+    free(cbuffer);
+}
+
+void circular_buffer_push(CircularBuffer_t *cbuffer, float data)
+{
+    cbuffer->buffer_data[cbuffer->head] = data;
+    cbuffer->head                       = (++cbuffer->head) % cbuffer->max_size;
+    if (circular_buffer_is_full(cbuffer) == true)
+    {
+        cbuffer->tail = (++cbuffer->tail) % cbuffer->max_size;
+    }
+
+    // Implementation assumes head and tail are equal when full
+    if (cbuffer->head == cbuffer->tail)
+    {
+        cbuffer->isFull = true;
+    }
+}
+
+float circular_buffer_get_index(CircularBuffer_t *cbuffer, int index)
```

Comment:
should `index` be unsigned? Don't think negative values are valid

Reply:
Changed it to be `uint8_t`.

Reply:
What about making it `size_t` since then it's the same range as the max size of the buffer?